### DISCUSSION TITLE
using lambda expression parameter forms is causing enhancement failures

### DISCUSSION THREAD

**Discussion by YANGJINJUE** [apache/skywalking#12008] [Q&A]
If using lambda expression parameter forms is causing enhancement failures, and I'm currently using SkyWalking version 8.1.0, is it possible to fix this issue without upgrading the version?

Example:
```public class ConsumeMessageConcurrentlyInstrumentation extends ClassInstanceMethodsEnhancePluginDefine {
    private static final String ENHANCE_CLASS = "org.apache.rocketmq.client.consumer.listener.MessageListenerConcurrently";
    private static final String CONSUMER_MESSAGE_METHOD = "consumeMessage";
    private static final String INTERCEPTOR_CLASS = "org.apache.skywalking.apm.plugin.rocketMQ.v4.MessageConcurrentlyConsumeInterceptor";

    @Override
    public ConstructorInterceptPoint[] getConstructorsInterceptPoints() {
        return new ConstructorInterceptPoint[0];
    }

    @Override
    public InstanceMethodsInterceptPoint[] getInstanceMethodsInterceptPoints() {
        return new InstanceMethodsInterceptPoint[] {
            new InstanceMethodsInterceptPoint() {
                @Override
                public ElementMatcher<MethodDescription> getMethodsMatcher() {
                    return named(CONSUMER_MESSAGE_METHOD);
                }

                @Override
                public String getMethodsInterceptor() {
                    return INTERCEPTOR_CLASS;
                }

                @Override
                public boolean isOverrideArgs() {
                    return false;
                }
            }
        };
    }

    @Override
    protected ClassMatch enhanceClass() {
        return HierarchyMatch.byHierarchyMatch(new String[] {ENHANCE_CLASS});
    }
}
```
My Code:
```
consumer.registerMessageListener((MessageListenerConcurrently)(msgs, context) -> {
          //TODO
    }
```

**Comment by wu-sheng** [comment]
Lambda itself is just syntactic sugar only. It always targets a class/interface, but as an anonymous implementation. Nothing more.

The interceptor points in the codes are just targeting the class(MessageListenerConcurrently#consumeMessage). If your codes are not consuming from there, it is not traced. That could be a reason.

> is it possible to fix this issue without upgrading the version?

I don't know what you mean by this, actually.

**Reply by YANGJINJUE** [reply]
```
consumer.registerMessageListener(new MessageListenerConcurrently() {
                @Override
                public ConsumeConcurrentlyStatus consumeMessage(List<MessageExt> list, ConsumeConcurrentlyContext consumeConcurrentlyContext) {
            });
```
The above approach works fine; MessageListenerConcurrently instanceof EnhancedInstance returns true. However, if lambda expression parameters are causing enhancement failures, MessageListenerConcurrently instanceof EnhancedInstance returns false.

**Reply by wu-sheng** [reply]
Lambda class may not be able to enhance, you need to cover it to a wrapper, a.k.a. the old lambda if you enhanced lambda to do the work.
If you know how to write a plugin, you should be able to understand what I mean.

**Reply by YANGJINJUE** [reply]
```
public class ConsumerStartInstrumentation extends ClassInstanceMethodsEnhancePluginDefine {
    private static final String PUSH_ENHANCE_CLASS = "org.apache.rocketmq.client.consumer.MQPushConsumer";
    private static final String REGISTER_METHOD = "registerMessageListener";
    private static final String REGISTER_INTERCEPTOR_CLASS = "org.apache.skywalking.apm.plugin.rocketMQ.v4.ConsumerRegisterListenerInterceptor";

    @Override
    public ConstructorInterceptPoint[] getConstructorsInterceptPoints() {
        return new ConstructorInterceptPoint[0];
    }

    @Override
    public InstanceMethodsInterceptPoint[] getInstanceMethodsInterceptPoints() {
        return new InstanceMethodsInterceptPoint[]{
                new InstanceMethodsInterceptPoint() {
                    @Override
                    public ElementMatcher<MethodDescription> getMethodsMatcher() {
                        return named(REGISTER_METHOD);
                    }

                    @Override
                    public String getMethodsInterceptor() {
                        return REGISTER_INTERCEPTOR_CLASS;
                    }

                    @Override
                    public boolean isOverrideArgs() {
                        return false;
                    }
                }
        };
    }

    @Override
    protected ClassMatch enhanceClass() {
        return byHierarchyMatch(new String[]{PUSH_ENHANCE_CLASS});
    }
}
```
```
public class ConsumerRegisterListenerInterceptor implements InstanceMethodsAroundInterceptor {

    @Override
    public final void beforeMethod(EnhancedInstance objInst, Method method, Object[] allArguments,
                                   Class<?>[] argumentsTypes, MethodInterceptResult result) throws Throwable {
        DefaultMQPushConsumer consumer = (DefaultMQPushConsumer) objInst;
        ConsumerEnhanceRequiredInfo requiredInfo = new ConsumerEnhanceRequiredInfo(consumer.getNamesrvAddr(), consumer.getConsumerGroup());
        Object argument0 = allArguments[0];
        if (argument0 instanceof EnhancedInstance) {
            EnhancedInstance messageListenerEnhanced = (EnhancedInstance) argument0;
            messageListenerEnhanced.setSkyWalkingDynamicField(requiredInfo);
        }
    }

    @Override
    public Object afterMethod(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes,
                              Object ret) throws Throwable {
        return ret;
    }

    @Override
    public void handleMethodException(EnhancedInstance objInst, Method method, Object[] allArguments,
                                      Class<?>[] argumentsTypes, Throwable t) {
        ContextManager.activeSpan().errorOccurred().log(t);
    }
}
```
argument0 instanceof EnhancedInstance return false

**Reply by wu-sheng** [reply]
Are you reading my reply? I am confused about your point of posting the codes. If the Lambda class is not instrumented, it is not `instanceof` EnhancedInstance. There is nothing relative with your posted codes.

**Comment by geek-frio** [comment]
@wu-sheng 这里的问题应该是skywalking的AgentBuilder在初始化的时候并未指定: AgentBuilder.LambdaInstrumentationStrategy.ENABLED。

在skywalking的插件里面针对RocketMQ消费消息的拦截实现，是对MessageListenerConcurrently的consumeMessage的增强。对于使用rocketmq进行业务开发的人员来说，进行注册消费的MessageListenerConcurrently的时候，很多都是使用Lambda的方式进行注入MessageListenerConcurrently的匿名类的，这就会导致skywalking对于这种情况的trace的消费链路断掉，出现没有消费trace的情况。我们刚才尝试再skywalking-agent里面去打开这个AgentBuilder.LambdaInstrumentationStrategy.ENABLED这个配置，但是出现了其它未预期的报错。

The issue here seems to be that the AgentBuilder in SkyWalking is not initialized with AgentBuilder.LambdaInstrumentationStrategy.ENABLED specified.

In SkyWalking's plugins, the interception implementation for consuming messages in RocketMQ focuses on enhancing the consumeMessage method of MessageListenerConcurrently. For developers using RocketMQ for their business, when registering a consumer with MessageListenerConcurrently, they often use Lambda expressions to inject anonymous classes implementing MessageListenerConcurrently. This can lead to a situation where SkyWalking's tracing of the consumption chain is disrupted, resulting in the absence of consumption traces. We have tried to enable AgentBuilder.LambdaInstrumentationStrategy.ENABLED in the SkyWalking agent, but encountered unexpected errors.

**Reply by wu-sheng** [reply]
I am not going to discuss in Chinese. It is not friendly for others.
Keep English on GitHub, or use Chinese on slack CN channel.

**Reply by wu-sheng** [reply]
> The issue here seems to be that the AgentBuilder in SkyWalking is not initialized with AgentBuilder.LambdaInstrumentationStrategy.ENABLED specified.
In SkyWalking's plugins, the interception implementation for consuming messages in RocketMQ focuses on enhancing the consumeMessage method of MessageListenerConcurrently. For developers using RocketMQ for their business, when registering a consumer with MessageListenerConcurrently, they often use Lambda expressions to inject anonymous classes implementing MessageListenerConcurrently. This can lead to a situation where SkyWalking's tracing of the consumption chain is disrupted, resulting in the absence of consumption traces. We have tried to enable AgentBuilder.LambdaInstrumentationStrategy.ENABLED in the SkyWalking agent, but encountered unexpected errors.

I would not recommend to open enable this, I don't know the impact. Usually, as I mentioned, you could use a wrapper to make this work.

**Reply by geek-frio** [reply]
Using a Wrapper is a direct approach, but it cannot solve the problem that the behavior of other upper-level business development  programmer using the RocketMQ client package cannot be directly constrained. What I want to know is, do you understand the reason why lambda cannot be enabled?

**Reply by wu-sheng** [reply]
No, I don't. We don't want to touch on that. Lambda is tricky from the agent side.

**Reply by wu-sheng** [reply]
For business things, don't try to deal with that through your way. It is not stable.
Use the toolkit lib with your application codes, it is safer and more stable.
